### PR TITLE
Do not add `;` to expected tokens list when it's wrong

### LINKED ISSUE #87647: Parser incorrectly suggests adding semicolon after function in impl
This also ICEs, but that's a separate issue: #87635. The MCVE here is also from that issue.

I think this issue is also caused by #87436, based on looking at the changes in the test output from that PR. That PR seems to have changed the parser to emit an ``expected `;`, found ...`` error regardless of what token was actually expected.

### Code

[Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=c8d3f3351aa7811e270905d8c5a76475)

```rust
struct Foo {}

impl Foo {
    pub fn bar()
}

fn main() {}
```

### Nightly error

```
error: expected `;`, found `}`
 --> src/main.rs:4:17
  |
4 |     pub fn bar()
  |                 ^ help: add `;` here
5 | }
  | - unexpected token

thread 'rustc' panicked at 'internal error: entered unreachable code', compiler/rustc_parse/src/parser/item.rs:1723:17
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.56.0-nightly (492723897 2021-07-29) running on x86_64-unknown-linux-gnu

note: compiler flags: -C embed-bitcode=no -C codegen-units=1 -C debuginfo=2 --crate-type bin

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
end of query stack
error: could not compile `playground` due to previous error
```

### Beta error

```
error: expected one of `->`, `;`, `where`, or `{`, found `}`
 --> src/main.rs:5:1
  |
4 |     pub fn bar()
  |            ---  - expected one of `->`, `;`, `where`, or `{`
  |            |
  |            while parsing this `fn`
5 | }
  | ^ unexpected token

error: associated function in `impl` without body
 --> src/main.rs:4:5
  |
4 |     pub fn bar()
  |     ^^^^^^^^^^^-
  |                |
  |                help: provide a definition for the function: `{ <body> }`

error: could not compile `playground` due to 2 previous errors
```

### Meta

Regressed in nightly 1.56.0.

### PR DESCRIPTION
**Pull request by notriddle**
There's a few spots where semicolons are checked for to do error recovery, and should not be suggested (or checked for other stuff).

Fixes #87647

### CODE REVIEW

**Comment by michaelwoerister** [comment]
r? rust-lang/diagnostics

File: compiler/rustc_parse/src/parser/item.rs [hunk 1]
```diff
@@ -1700,23 +1708,33 @@ impl<'a> Parser<'a> {
...
 
+/// Parsing configuration for functions.
+/// This include the edition-specific name requirements, plus information on whether the
+/// function is allowed to go without a body.
+#[derive(Clone, Copy)]
+pub(crate) struct FnParseMode {
+    pub req_name: ReqName,
+    pub req_body: bool,
+}
+
```
Comment (estebank) [inline compiler/rustc_parse/src/parser/item.rs:1719]: Could you extend the docstring ("what, why and where" this affects the parser), add a docstring to each field so it's more immediately understood what each of them do.

Comment (notriddle) [reply compiler/rustc_parse/src/parser/item.rs]: Okay, I've added more comments to it.

**Pull request merged**